### DISCUSSION TITLE
OWL vs ShEx?

### DISCUSSION THREAD

**Discussion by photocyte** [Bioprotocols/container-ontology#25] [General]
I've been exploring OWL a bit more. Namely, trying to understand it from the perspective of Wikidata, since that was my introduction to Semantic Web technologies. 

[This paper](https://bmcbiol.biomedcentral.com/articles/10.1186/s12915-020-00940-y), which describes a workflow for adding COVID-19 data to Wikidata, talks a lot about a technology known as ShEx, and its editors like [YaShe](https://www.weso.es/YASHE/) and [sheXer](https://github.com/DaniFdezAlvarez/shexer), & the process of iterative development using those tools to evaluate the structure of properties that already existed on Wikidata, & help plan out improvements to properties. [This snippet](http://shex.io/shex-primer/#relationship) of the ShEx documentation describes the cross comparison between OWL and ShEx. 

I definitely don't understand the subtleties of OWL/RDF/ShEx, but I guess I'm wondering if OWL is the right tool vs ShEx for the container ontology? I'm mainly coming at this from the "property provenance" angle (i.e., if a plate height is explicitly added for a given plate item, did that come from a vendor datasheet, a physical measurement by the contributor, other sources i.e. Strateos or Opentrons libraries?), since I think that is critical feature for a multi-contributor ontology, & it sounded from the recent container ontology meeting like the OWL mechanisms for that were not fully available or not very user friendly?

In contrast, every "triple" on Wikidata has the ability for provenance via the "reference" annotation (i.e. see [here](https://www.wikidata.org/wiki/Q82069695)), and furthermore have the ability for [qualifiers](https://www.wikidata.org/wiki/Help:Qualifiers) to be added to the given triple. There are Wikidata properties like "Instance of", https://www.wikidata.org/wiki/Property:P31 , which implement a form of inheritance, and I think there may be a way to use a qualifer to indicate negation of the triple. [This discussion](https://www.wikidata.org/wiki/Wikidata:Requests_for_comment/How_to_capture_negative_results_in_Wikidata%3F) suggests there was not at consensus in ~2019 on how to negate triples at least, but perhaps a consensus has been reached since then...

**Comment by rpgoldman** [comment]
> I've been exploring OWL a bit more. Namely, trying to understand it from the perspective of Wikidata, since that was my introduction to Semantic Web technologies.
> 
> [This paper](https://bmcbiol.biomedcentral.com/articles/10.1186/s12915-020-00940-y), which describes a workflow for adding COVID-19 data to Wikidata, talks a lot about a technology known as ShEx, and its editors like [YaShe](https://www.weso.es/YASHE/) and [sheXer](https://github.com/DaniFdezAlvarez/shexer), & the process of iterative development using those tools to evaluate the structure of properties that already existed on Wikidata, & help plan out improvements to properties. [This snippet](http://shex.io/shex-primer/#relationship) of the ShEx documentation describes the cross comparison between OWL and ShEx.

I'm not sure how to evaluate that discussion of the relationship between OWL and ShEX. Reading over it suggests that ShEX might have more in common with [SHACL](https://www.w3.org/TR/shacl/).  SHACL is a notation that lets one express constraints on the "shape" of RDF graphs.  TBH, I couldn't make sense out of this claim:

> > A ShEx schema describes something in the world, where that "something" happens to be the set of abstract-syntactic components that comprise an RDF graph. However, such an interpretation is of no use in practical terms.

> I definitely don't understand the subtleties of OWL/RDF/ShEx, but I guess I'm wondering if OWL is the right tool vs ShEx for the container ontology? I'm mainly coming at this from the "property provenance" angle (i.e., if a plate height is explicitly added for a given plate item, did that come from a vendor datasheet, a physical measurement by the contributor, other sources i.e. Strateos or Opentrons libraries?), since I think that is critical feature for a multi-contributor ontology, & it sounded from the recent container ontology meeting like the OWL mechanisms for that were not fully available or not very user friendly?

I think that OWL is a good tool for the container ontology because it specifies a method (a semantics and algorithms that match that semantics) for determining whether one entity matches a description that we can use to answer questions like "is this container suitable to fill *X* role in this protocol," and there are tools that we can get off the shelf to do that task.  That is different from Wikidata, which doesn't have a semantics for "instance of," and so does not dictate what it means for one entity to be an instance of a class.  That's not surprising, since OWL aims at supplying sound and efficient reasoning about descriptions, inheritance, etc.  On the one hand, that means that OWL can do some things with formally defined concepts that I don't believe Wikidata's scheme can do. On the other hand, of course, this means that OWL *can't* do things with informally defined concepts ("natural kinds") because it conflicts with the monotonic semantics of OWL.

I think the semantics and inference of OWL are just fine for the container ontology, given its primary use case, and given that the subject of this ontology is human-made artifacts, rather than informal entities. But as I have said before, I *don't* think OWL would be great for describing, for example, organisms.

As for provenance, I don't think OWL, ShEx, or any other RDF dialect differ in their ability to have their provenance recorded.  As long as they have RDF semantics, they will correspond to triples and can be annotated in the same ways.  I'm not sure why Wikidata prefers the "reference" annotation over using the provenance ontology, but I'm really almost completely ignorant about Wikidata.  Note that annotating triples requires somehow "reifying" them (turning them into things), since a triple is not a thing, it's a bundle of three things. So it can't have properties itself, because of the inherent limitations of RDF. So we have to create a node in the graph that is the "name" of the triple, and we can assign properties to that node.

I imagine it should be possible to provide a translation between the container ontology and information in Wikidata.  If there is useful information in Wikidata, it would be very handy to be able to extract it.

One could also make the contents of the container ontology available to code that works on Wikidata.

**Comment by rpgoldman** [comment]
Just checked and there is [a page comparing ShEx and SHACL](https://www.w3.org/2014/data-shapes/wiki/SHACL-ShEx-Comparison) which must be sorta authoritative, or at least as authoritative as the W3C gets.

**Comment by photocyte** [comment]
Thanks Robert, I trust your judgement on the technical comparisons & call to stick with OWL, although I'm glad that the comparisons were also not so clearly explained for you - seems the comparisons are just tricky. A few more resources to share to on my overall thinking, not sure if it would be useful to other folks who might find this thread...

>  I'm not sure why Wikidata prefers the "reference" annotation over using the provenance ontology,

I think that may have to do with fact that the interoperability of Wikidata w/ RDF & SPARQL was a comparatively late design decision. And for something like "references" or "citations" on #citationneeded Wikipedia / Wikimedia, I presume there was no way back to use more of the semantic provenance frameworks. See below reference for that story:

Malyshev S, Krötzsch M, González L, Gonsior J, Bielefeldt A. 2018. Getting the Most Out of Wikidata: Semantic Technology Usage in Wikipedia’s Knowledge Graph In: Vrandečić D, Bontcheva K, Suárez-Figueroa MC, Presutti V, Celino I, Sabou M, Kaffee L-A, Simperl E, editors. The Semantic Web – ISWC 2018, Lecture Notes in Computer Science. Cham: Springer International Publishing. pp. 376–394. doi:[10.1007/978-3-030-00668-6_23](https://doi.org/10.1007/978-3-030-00668-6_23) 

It's *possible* the RDF dump of Wikidata is using the provenance functionality, but I'd sort of doubt it, and haven't looked into it.

> If there is useful information in Wikidata, it would be very handy to be able to extract it.

I don't think there is any data relevant to the container-ontology / equipment-ontology currently in Wikidata. 

> I imagine it should be possible to provide a translation between the container ontology and information in Wikidata.  ...
> One could also make the contents of the container ontology available to code that works on Wikidata.

But, it is indeed possible there could be a bi-directional interoperability. The decision to try and develop that functionality really comes down on how broadly the ontology's are expected to be edited/viewed (how "user friendly", query-able in different forms). 

If the documentation for contributing via the OWL tooling is clearly described, presumably the Git type workflow is not too big of a barrier for contributors vs the Wikibase type interactions, for the subset of folks who might be interested in contributing to the PAML-relevant ontologies. 

On broader RDF, semantic web stuff, etc. I was struck by the "Open world assumption", and in fact how RDF can't really capture negative data (so it's not just OWL that is constrained?) The below recent papers were helpful for grounding my semantic experience from Wikidata in the broader "Open world assumption" & it's limitations that might be cropping up in PAML... Maybe useful for someone.

Arnaout H, Razniewski S, Weikum G, Pan JZ. 2021. Negative Knowledge for Open-world WikidataCompanion Proceedings of the Web Conference 2021. Presented at the WWW ’21: The Web Conference 2021. Ljubljana Slovenia: ACM. pp. 544–551. doi:[10.1145/3442442.3452339](https://doi.org/10.1145/3442442.3452339)

Straccia U, Casini G. 2022. A Minimal Deductive System for RDFS with Negative Statements. arXiv:220213750 [cs]. (http://arxiv.org/abs/2202.13750)
